From 56dca8fedfa12218a0172f6a79f7b586f66166c8 Mon Sep 17 00:00:00 2001
From: Ikey Doherty <ikey@solus-project.com>
Date: Mon, 5 Dec 2016 02:26:08 +0000
Subject: [PATCH] Add new "-t" option for reproducible builds

This parameter will be passed to us from solbuild, with the timestamp
for the last tag in the package. This ensures that unmodifified source
packages with the same tag will always share the same timestamp throughout.

Signed-off-by: Ikey Doherty <ikey@solus-project.com>
---
 ypkg2/main.py     | 4 ++++
 ypkg2/metadata.py | 7 +++++++
 2 files changed, 11 insertions(+)

diff --git a/ypkg2/main.py b/ypkg2/main.py
index 2f028e5..ff62358 100644
--- a/ypkg2/main.py
+++ b/ypkg2/main.py
@@ -51,6 +51,8 @@ def main():
                         action="store_true")
     parser.add_argument("-v", "--version", action="store_true",
                         help="Show version information and exit")
+    parser.add_argument("-t", "--timestamp", help="Unix timestamp for build",
+                        type=int, default=-1)
     parser.add_argument("-D", "--output-dir", type=str,
                         help="Set the output directory for resulting files")
     # Main file
@@ -66,6 +68,8 @@ def main():
     # Show version
     if args.version:
         show_version()
+    if args.timestamp > 0:
+        metadata.history_timestamp = args.timestamp
 
     if args.output_dir:
         od = args.output_dir
diff --git a/ypkg2/metadata.py b/ypkg2/metadata.py
index 996b882..cf295e9 100644
--- a/ypkg2/metadata.py
+++ b/ypkg2/metadata.py
@@ -75,6 +75,13 @@ def initialize_timestamp(spec):
     global history_timestamp
     global fallback_timestamp
 
+    # If it was already set, just skip it.
+    if history_timestamp is not None:
+        dt = datetime.datetime.fromtimestamp(history_timestamp)
+        history_date = dt.strftime("%Y-%m-%d")
+        fallback_timestamp = history_timestamp
+        return
+
     dt = datetime.datetime.utcnow()
     s = dt.strftime("%Y-%m-%d")
     fallback_timestamp = utc_date_for_date_only(s)
-- 
2.11.0

